### ISSUE TITLE
Code Review

### ISSUE THREAD

**Issue by MohamedBassem** [AAbouZaid/golang-playground#2]
Not sure what's the best way to review an existing code so I'll do it here:

1. First of all, and the most important one, this code is racy. You are passing the same map to multiple goroutines and decoding a JSON into the map in each, they will override each other.
2. Also, Go maps are not thread safe. You are concurrently reading/writing values to the same map which is wrong. If you run you code with [race detector](https://blog.golang.org/race-detector) enabled, your code will probably crash. Which takes me to the next point.
3. It doesn't seem like you need to synchronize the goroutines at all. Why not pass each one its own instance of the map? Even better, change `gatherFunction` to be a function instead of a method and change `getJSON` to be the one returning the map. To be honest, you don't need to attach your functions to any objects at all. I'd prefer passing the map the functions that need it instead.
4. Passing maps everywhere makes it very hard to follow the code. Something like the `fixedMap` variable. It should have been a struct given that you already know all the keys. As long as you know the structure of your object, use structs instead of map to make it easier to read and understand.
5. It's fine if you implement your own `inArray` function. It's trivial and you don't want to import a whole new package to implement it for you.
6. If I understand you comment on top of the `defer recover()` correctly, it's because you are using `log.Fatalf`. It logs and then terminates your program. So if any of the servers is down, whenever its goroutine reaches the `log.Fatalf`, the code will terminate. To fix that:
  a. You don't even need the `defer recover()`, I don't see anything `panic`-ing here.
  b. If you don't want your program to crash, use `log.Printf` instead, or if you want to handle the error after you are done, pass an error channel to your goroutines.

**Comment by MohamedBassem** [comment]
Actually, I'm confused with the way you are sometimes pass a pointer receiver and sometimes a value receiver when the receiver is a map. Honestly, right now I don't even know if it's racy or not. But in both cases, don't do it this way.

**Comment by aabouzaid** [comment]
Thanks a lot Mohamed for your review :D
And sorry for that code not really clear ... that what you have when system engineer code in a new language :D

> 1. First of all, and the most important one, this code is racy. You are passing the same map to multiple goroutines and decoding a JSON into the map in each, they will override each other.
There are 2 maps in the code.

First one is just a fixed config, so no overwritten happens at any level.

Second one uses a copy of the map, where it happened once per server ...
so each time get json from a url, it will write to "metrics" to put its data.

that's why the output is correct at the end.

> 2. Also, Go maps are not thread safe. You are concurrently reading/writing values to the same map which is wrong. If you run you code with [race detector](https://blog.golang.org/race-detector) enabled, your code will probably crash. Which takes me to the next point.

Hmmm I have no idea about "race detector" so I will read about it :D 
But the code is working with `-race` and it didn't crash (I still have no idea what is that, I still need to read about it 😂 ).

> 3. It doesn't seem like you need to synchronize the goroutines at all. Why not pass each one its own instance of the map? Even better, change `gatherFunction` to be a function instead of a method and change `getJSON` to be the one returning the map. To be honest, you don't need to attach your functions to any objects at all. I'd prefer passing the map the functions that need it instead.

Yeah, I think so too :D
I realized that once I need to pass a new var to last method in the chain ... where I needed to pass that var through the whole chain :D

> 4. Passing maps everywhere makes it very hard to follow the code. Something like the `fixedMap` variable. It should have been a struct given that you already know all the keys. As long as you know the structure of your object, use structs instead of map to make it easier to read and understand.

The keys in json are not the same, they could be anything, that's why I used a map for that.

> 5. It's fine if you implement your own `inArray` function. It's trivial and you don't want to import a whole new package to implement it for you.

I'm just not sure if there another way to do it, which not necessary the same :D 

> 6. If I understand you comment on top of the `defer recover()` correctly, it's because you are using `log.Fatalf`. It logs and then terminates your program. So if any of the servers is down, whenever its goroutine reaches the `log.Fatalf`, the code will terminate. To fix that:
  a. You don't even need the `defer recover()`, I don't see anything `panic`-ing here.
  b. If you don't want your program to crash, use `log.Printf` instead, or if you want to handle the error after you are done, pass an error channel to your goroutines.

I think `Fatalf` shouldn't be used, but here is what was my question:


if we have 2 servers ... ok we have 4 cases:

- both servers are working (return 200).
= in this case everything fine.

- 1st is working, 2nd not.
= in this case the output of 1st server printed, and error from 2nd server printed.

- 1st is not working, 2nd is.
= in this case the error from 1st server printed and the output of 2nd server printed

- both are not working.
= in this case the error of 1st server "only" is printed ... and here is my question why that happened.


Thanks a lot for your comments, they are very helpful \o/

**Comment by aabouzaid** [comment]
@MohamedBassem Later I found what do you mean by race conditions :grin: 
and I did take care of it in the [real version](https://github.com/AAbouZaid/telegraf-flume-plugin/blob/master/flume/flume.go) of the plugin.

Thanks a lot again for your feedback :tada:
